R/createdDdl.R

```diff
@@ -0,0 +1,140 @@
+
+#' Create the OHDSI-SQL Common Data Model DDL code
+#'
+#' The createDdl, createForeignKeys, and createPrimaryKeys functions each return a character string
+#' containing their respective DDL SQL code in OHDSQL dialect for a specific CDM version.
+#' The SQL they generate needs to be rendered and translated before it can be executed.
+#'
+#' The DDL SQL code is created from a two csv files that detail the OMOP CDM Specifications.
+#' These files also form the basis of the CDM documentation and the Data Quality
+#' Dashboard.
+#'
+#' @param cdmVersion The version of the CDM you are creating, e.g. 5.3, 5.4
+#' @return A character string containing the OHDSQL DDL
+#' @export
+#' @examples
+#' ddl <- createDdl("5.4")
+#' pk <- createPrimaryKeys("5.4")
+#' fk <- createForeignKeys("5.4")
+createDdl <- function(cdmVersion){
+
+  # argument checks
+  stopifnot(is.character(cdmVersion), length(cdmVersion) == 1, cdmVersion %in% listSupportedVersions())
+
+  cdmTableCsvLoc <- system.file(file.path("csv", paste0("OMOP_CDMv", cdmVersion, "_Table_Level.csv")), package = "CommonDataModel", mustWork = TRUE)
+  cdmFieldCsvLoc <- system.file(file.path("csv", paste0("OMOP_CDMv", cdmVersion, "_Field_Level.csv")), package = "CommonDataModel", mustWork = TRUE)
+
+  tableSpecs <- read.csv(cdmTableCsvLoc, stringsAsFactors = FALSE)
+  cdmSpecs <- read.csv(cdmFieldCsvLoc, stringsAsFactors = FALSE)
+
+  tableList <- tableSpecs$cdmTableName
+
+  sql_result <- c()
+  sql_result <- c(paste0("--@targetDialect CDM DDL Specification for OMOP Common Data Model ", cdmVersion))
+  for (tableName in tableList){
+    fields <- subset(cdmSpecs, cdmTableName == tableName)
+    fieldNames <- fields$cdmFieldName
+
+    if ('person_id' %in% fieldNames){
+      query <- "\n\n--HINT DISTRIBUTE ON KEY (person_id)\n"
+    } else {
+      query <- "\n\n--HINT DISTRIBUTE ON RANDOM\n"
+    }
+
+    sql_result <- c(sql_result, query, paste0("CREATE TABLE @cdmDatabaseSchema.", tableName, " ("))
+
+    n_fields <- length(fieldNames)
+    for(fieldName in fieldNames) {
+
+      if (subset(fields, cdmFieldName == fieldName, isRequired) == "Yes") {
+        nullable_sql <- (" NOT NULL")
+      } else {
+        nullable_sql <- (" NULL")
+      }
+
+      if (fieldName == fieldNames[[n_fields]]) {
+        closing_sql <- (" );")
+      } else {
+        closing_sql <- (",")
+      }
+
+      if (fieldName=="offset") {
+        field <- paste0('"',fieldName,'"')
+      } else {
+        field <- fieldName
+      }
+      fieldSql <- paste0("\n\t\t\t",
+                         field," ",
+                         subset(fields, cdmFieldName == fieldName, cdmDatatype),
+                         nullable_sql,
+                         closing_sql)
+      sql_result <- c(sql_result, fieldSql)
+    }
+    sql_result <- c(sql_result, "")
+  }
+  return(paste0(sql_result, collapse = ""))
+}
+
+
+#' @describeIn createDdl createPrimaryKeys asdfladsfj
+#' @export
+createPrimaryKeys <- function(cdmVersion){
```

Comment:
I moved the "create" functions into a single file and used a single documentation page for them.